### PR TITLE
Live view panel

### PR DESCRIPTION
**Pull request by RandomShaper**
On top of some non-breaking changes to `Image` and a cross-platform implementation of shared memory, this is a dock that shows a small view of the running game.

~~After a lot of thought and given the current limitations of the Godot editor UI, I've concluded that making it a dock was the best option.~~

~~To avoid it taking editor space when it isn't in use, in the default layout, it will be beneath the file system tab. It also has settings (enabled by default) to auto-switch to it and back to whatever other tab was active before running the game.~~

Also by default, it's enabled the option that makes it stand by while the editor is not focused. That is good because it prevents it from stealing performance from the running game and anyway you don't need to look at a small view in the editor while the game window is the focused one. Nonetheless any user wanting to do so will be able to.

It has a maximum FPS limit to keep it's load light, set to 15 by default. Someone using this on a powerful enough system will be able to raise it.

Furthermore, whenever it's not visible ~~(because another dock in the same tab group is active)~~, it is effectively disabled.

Given that the framebuffer data is transferred via shared memory, this only works when both the game and the editor are running on the same host.

**~~I've not been able to confirm image capture is working on Mac. If anyone could test and report, that'd be great.~~** **UPDATE:** It seems it does!

**UPDATE:**
~~I'm marking this as _work-in-progress_ because I will add some more pieces of functionality to make it more useful. However, the trickiest part is done and testing is welcome.~~

**UPDATE 2:**
No longer work-in-progress. I've moved it to the output dock. Additional functionality will come in future PRs.

image
image

### CODE REVIEW

File: core/os/shared_memory.cpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
+/*************************************************************************/
+/*  shared_memory.cpp                                                */
```
Comment (akien-mga) [inline core/os/shared_memory.cpp]: End `*/` is not aligned.

File: core/os/shared_memory.h [hunk 2]
```diff
@@ -0,0 +1,103 @@
+/*************************************************************************/
+/*  shared_memory.h                                                  */
```
Comment (akien-mga) [inline core/os/shared_memory.h]: Same here.

File: platform/windows/shared_memory_windows.h [hunk 3]
```diff
@@ -0,0 +1,67 @@
+/*************************************************************************/
+/*  shared_memory_posix.h                                            */
```
Comment (akien-mga) [inline platform/windows/shared_memory_windows.h]: Should be `_windows.h`, and end tag also misaligned.

File: core/os/shared_memory.cpp (hunk 1 above)
Comment (RandomShaper) [reply core/os/shared_memory.cpp]: Fixed!

File: core/os/shared_memory.h (hunk 2 above)
Comment (RandomShaper) [reply core/os/shared_memory.h]: Fixed!

File: platform/windows/shared_memory_windows.h (hunk 3 above)
Comment (RandomShaper) [reply platform/windows/shared_memory_windows.h]: Fixed!

Comment (akien-mga) [reply platform/windows/shared_memory_windows.h]: Actually this file duplicates `drivers/windows/shared_memory_windows.h`.

Comment (RandomShaper) [reply platform/windows/shared_memory_windows.h]: Oops. You're right. Let me check...

File: drivers/windows/shared_memory_windows.h [hunk 4]
```diff
@@ -0,0 +1,80 @@
+/*************************************************************************/
+/*  shared_memory_posix.h                                                */
```
Comment (akien-mga) [inline drivers/windows/shared_memory_windows.h]: windows

File: platform/windows/shared_memory_windows.h (hunk 3 above)
Comment (RandomShaper) [reply platform/windows/shared_memory_windows.h]: Fixed!

File: drivers/windows/shared_memory_windows.h (hunk 4 above)
Comment (RandomShaper) [reply drivers/windows/shared_memory_windows.h]: Fixed!

File: platform/x11/detect.py [hunk 5]
```diff
@@ -318,7 +318,7 @@ def configure(env):
 
     env.Prepend(CPPPATH=['#platform/x11'])
     env.Append(CPPFLAGS=['-DX11_ENABLED', '-DUNIX_ENABLED', '-DOPENGL_ENABLED', '-DGLES_ENABLED'])
-    env.Append(LIBS=['GL', 'pthread'])
+    env.Append(LIBS=['GL', 'pthread', 'rt'])
```
Comment (RandomShaper) [inline platform/x11/detect.py]: @akien-mga, @hpvb, I think this has implications for CI and build containers.

**Comment by aaronfranke** [comment]
Doesn't this then require that the game be running? How useful is this exactly if by running the game it brings itself to focus as a separate window anyway?

**Comment by RandomShaper** [comment]
The main idea is using it for live scene editing and debugging without having to switch back and forth between the game and the editor.
It's not a one-fits-all solution, neither regarding use cases nor projects, but it was requested by some people to a extent that it was added to the roadmap.

**Comment by Zireael07** [comment]
Is this for 2D only or 3D, too?

**Comment by santouits** [comment]
beautiful code but I don't get why is this needed. Is there anywhere any discussion about it to read?

**Comment by RandomShaper** [comment]
@Zireael07, it's for both. It mirrors anything displayed in the root (builtin) viewport.

@santouits, you're right in that it's not needed, but that doesn't mean users cannot benefit from it.

I'm not sure there's a thread somewhere about it. I just know several users requested it during spoken conversations.

Then we talked about it during the Godot Sprint and decided to include it in the roadmap.

Thanks for your words about the code.

**Comment by AinaSG** [comment]
Hey @RandomShaper, I compiled it in .a Mac and I see an static image after hiting play and closing the game, I don't know if this the expected behaviour :) 
Screen Shot 2019-05-24 at 16 11 18

**Comment by Zireael07** [comment]
If I'm reading it correctly, game must be running for it to work, so if you close it, it won't work.

**Comment by AinaSG** [comment]
Oh, right, it shows hen I give focus back to the editor, but is's till static I can't seem to be able to interact with it or the animations playing.

**Comment by RandomShaper** [comment]
@AinaSG, thank you very much for testing!

Yes, as it is currently this is quite basic, but actually being able to mirror the game image into the editor was the hardest part.

I think I'm marking this PR as work-in-progress, because I will do some improvements to make it more useful.

**Comment by Jummit** [comment]
Thank you for making this! I switch between the game and the editor very often, so this will be very useful.

**Comment by swarnimarun** [comment]
Cool feature. But what kind of performance repercussions will be there? I mean I saw from my quick look that it's locked at 15 fps by default, and I am assuming it needs to redraw the editor or can it just work by redrawing the dock separately in a different thread?
(Probably I should have seen the code rather than asking).

**Comment by RandomShaper** [comment]
@swarnimarun, the editor is redrawn on every frame. That's why by default it will be in stand by while the game window is focused, to avoid disrupting normal game performance.

When you focus the editor back, with the game still running, then a slight FPS drop is not as important since you are then focusing on fixing/editing/debugging.

**Comment by Calinou** [comment]
@AinaSG Does this occur if you disable fullscreen in the Project Settings? Maybe alt-tabbing out of a fullscreen project causes the renderer to stop outputting new frames (as the result won't be visible).

**Comment by swarnimarun** [comment]
@RandomShaper  That's nice, actually really awesome. 👍 
(At least I can't think of anything better unless we get detachable dock windows)

**Comment by paulhocker** [comment]
my $0.02

coming from a unity background, the integrated runner is actually more of a hindrance. I remember having to constantly resize windows or move them around. With Godot, all you have to do is alt-tab back to the engine when running the game. 

the fact that there is a fps throttle tells me that this solution might affect overall engine performance.

still, I guess I do not mind if this is some kind of plugin or module that you can add to the engine, but really do not want to see something like this added to the core.

**Comment by Jummit** [comment]
> coming from a unity background, the integrated runner is actually more of a hindrance. I remember having to constantly resize windows or move them around. With Godot, all you have to do is alt-tab back to the engine when running the game.

This is not a integrated game, just a preview of what is on screen.

**Comment by HeadClot** [comment]
Got a question - Is there a disable option for this feature?

**Comment by RandomShaper** [comment]
@HeadClot, currently the way of disabling it is implicit: if it's not the active tab, it's effectively stopped.

But I think I'll add more explicit control over that, along with the other enhancements.

**Comment by HeadClot** [comment]
@RandomShaper Thank you so much. :)

**Comment by paulhocker** [comment]
> 
> 
> > coming from a unity background, the integrated runner is actually more of a hindrance. I remember having to constantly resize windows or move them around. With Godot, all you have to do is alt-tab back to the engine when running the game.
> 
> This is not a integrated game, just a preview of what is on screen. (At least from what I understand)

the first sentence states:

"this is a dock that shows a small view of the running game"

is this PR intended to be added to the main engine for 3.2?

**Comment by Calinou** [comment]
@paulhocker They said it's not an integrated game because you can't control it from there, just look at it :slightly_smiling_face:

**Comment by Pr0methean** [comment]
This PR uses `shm_open` and `shm_unlink`, which don't exist on Android. This is why the Android tests failed on Travis.

**Comment by Pr0methean** [comment]
This PR uses shm_open and shm_unlink, which don't exist on Android. This is why the Android tests failed on Travis.

**Comment by RandomShaper** [comment]
You're right. I've hadn't time yet for fixing it.

**Comment by sparkart** [comment]
Clicking on the preview to freeze/unfreeze the preview would be cool.

File: core/os/shared_memory.h [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+	Also, bear in mind that resizing the shared memory block may be a heavy operation
+	on some platforms, so it's better to use this class for a sparingly resized block
+	or either overallocate and resize in big increments as needed.
+*/
+
+class SharedMemory {
+
+protected:
+	static const char *const ERR_STR_ALREADY_OPEN;
```
Comment (akien-mga) [inline core/os/shared_memory.h]: We know build with `-std=c++11` so I think you could move the definitions in the header (at last!).

Comment (RandomShaper) [reply core/os/shared_memory.h]: Done, as well as other C++11-related improvements.

File: core/image.cpp [hunk 7]
```diff
@@ -700,28 +700,57 @@ static void _scale_bilinear(const uint8_t *__restrict p_src, uint8_t *__restrict
...
 
-			for (uint32_t l = 0; l < CC; l++) {
+		switch (N) {
+			case 1: _scale_nearest_row_opt<uint8_t>(src_row, p_src_width, dst_row, p_dst_width); break;
+			case 2: _scale_nearest_row_opt<uint16_t>(src_row, p_src_width, dst_row, p_dst_width); break;
+			case 3: _scale_nearest_row_bytewise<3>(src_row, p_src_width, dst_row, p_dst_width); break;
+			case 4: _scale_nearest_row_opt<uint32_t>(src_row, p_src_width, dst_row, p_dst_width); break;
+			case 6: _scale_nearest_row_bytewise<6>(src_row, p_src_width, dst_row, p_dst_width); break;
+			case 8: _scale_nearest_row_opt<uint64_t>(src_row, p_src_width, dst_row, p_dst_width); break;
+		}
```
[the remark this answers is not in the document]
Comment (RandomShaper) [reply core/image.cpp]: `_scale_nearest`, being an implementation detail of `scale_nearest_raw` doesn't need to support anything aside from what the latter allows.

And the latter should allow any supported pixel size, so you've brought to my attention that 12 and 16 should be supported indeed.

I'm adding an error message in the case that someone uses an unknown pixel size.

**Comment by Anutrix** [comment]
Is this editor only or can it work as good minimap inside the game?

**Comment by RandomShaper** [comment]
Editor-only. The mechanism used to transfer the image from the game to the editor would be cumbersome for in-game use.

**Comment by akien-mga** [comment]
But unless I miss something, you don't need this feature at all in game. You already have a unified scene tree and you can use Viewports to achieve this kind of result, without having to bother with a cross-process communication channel.

**Comment by aaronfranke** [comment]
Closing because:

* This PR has not received any new commits for over a year, it needs rebasing.

* This feature needs a [proposal](https://github.com/godotengine/godot-proposals/issues) written to justify it.

  * I'm still not convinced that this feature is actually useful, or actually solving any particular problems. You can't control the game through this view. I think it's bloat, therefore I think it needs justification to be included in Godot.

  * I think this feature would be feasible to do in an add-on in the asset library. All it really needs is some way to grab another window and put it in a texture, then the add-on can display that texture anywhere, in a panel or dock, maybe even the main screen. If this would be included in Godot, a proposal needs to show that this isn't possible in a GDNative plugin (and if so, justify why we can't improve GDNative to allow this).

**Comment by starry-abyss** [comment]
> I'm still not convinced that this feature is actually useful
I was under impression OP was directly hired by Godot to implement this?

**Comment by RandomShaper** [comment]
> I was under impression OP was directly hired by Godot to implement this?

I wasn't. I did this because of an informal request for it, so to speak. It turned out to be a huge waste of time.

I hope that at some point shared memory is wanted for something else and at least that part can be used.

**Pull request closed without merging**